utils/runtime-fuzzer/src/gear_calls.rs

```diff
@@ -0,0 +1,359 @@
+        log::trace!(
+            "Random data after payload (upload_program) gen {}",
+            unstructured.len()
+        );
+        log::trace!("Payload (upload_program) length {:?}", payload.len());
+
+        let program_id = ProgramId::generate(CodeId::generate(&code), &salt);
+
+        log::trace!("Generated code for program id - {program_id}");
+
+        intermediate_data.uploaded_programs.push(program_id);
+
+        Ok(UploadProgramArgs((code, salt, payload, self.gas, self.value)).into())
+    }
+
+    fn unstructured_size_hint(&self) -> usize {
+        // 1024 KiB for payload and salt and 50 KiB for code.
+        1080 * 1024
+    }
+}
+
+/// Extrinsic generator that's capable of generating `SendMessage` calls.
+pub(crate) struct SendMessageGenerator {
+    pub gas: u64,
+    pub value: u128,
+    pub prepaid: bool,
+}
+
+impl ExtrinsicGenerator for SendMessageGenerator {
+    fn generate(
+        &self,
+        intermediate_data: &mut IntermediateData,
+        unstructured: &mut Unstructured,
+    ) -> Result<GearCall> {
+        let program_id = unstructured
+            .choose(&intermediate_data.uploaded_programs)
+            .copied()?;
+        let payload = arbitrary_payload(unstructured)?;
+        log::trace!(
+            "Random data after payload (send_message) gen {}",
+            unstructured.len()
+        );
+        log::trace!("Payload (send_message) length {:?}", payload.len());
+
+        Ok(SendMessageArgs((program_id, payload, self.gas, self.value, self.prepaid)).into())
+    }
+
+    fn unstructured_size_hint(&self) -> usize {
+        // 512 KiB for payload.
+        520 * 1024
+    }
+}
+
+/// Extrinsic generator that's capable of generating `SendReply` calls.
+pub(crate) struct SendReplyGenerator {
+    pub mailbox_provider: Box<dyn MailboxProvider>,
+
+    pub gas: u64,
+    pub value: u128,
+    pub prepaid: bool,
+}
+
+impl ExtrinsicGenerator for SendReplyGenerator {
+    fn generate(
+        &self,
+        _: &mut IntermediateData,
+        unstructured: &mut Unstructured,
+    ) -> Result<GearCall> {
+        log::trace!(
+            "Random data before payload (send_reply) gen {}",
+            unstructured.len()
+        );
+        let message_id =
+            arbitrary_message_id_from_mailbox(unstructured, self.mailbox_provider.as_ref())?;
+
+        let payload = arbitrary_payload(unstructured)?;
+        log::trace!(
+            "Random data after payload (send_reply) gen {}",
+            unstructured.len()
+        );
+        log::trace!("Payload (send_reply) length {:?}", payload.len());
+
+        Ok(SendReplyArgs((message_id, payload, self.gas, self.value, self.prepaid)).into())
+    }
+
+    fn unstructured_size_hint(&self) -> usize {
+        // 512 KiB for payload.
+        520 * 1024
+    }
+}
+
+/// Extrinsic generator that's capable of generating `ClaimValue` calls.
+pub(crate) struct ClaimValueGenerator {
+    pub mailbox_provider: Box<dyn MailboxProvider>,
```

Comment:
why there's a need for `MailboxProvider` trait object?

Reply:
We have all the logics related to runtime storage interactions in the `runtime` module, so this trait provides a level of abstraction over this logics

Reply:
I didn't get about abstractions over that logic. As far as I understand, the abstraction is enabled when the trait is used in both places - `runtime` and `gear_calls`, so then you depend on abstraction. But you don't need any abstraction here. If retrieve logic changes, you will change implementation accordingly without anything lost in build or perfomance.

I could say you need this abstraction from testing perspective, so you can provide your own `TestMailboxProvider`, but there's no even that.

So, are there still any reasons for including that trait object?

Reply:
Storage-related logics is relatively low-level, so we should keep it far from the higher-level fuzzing logics, doesn't it?

Reply:
you can use functions for making the code abstrac from "low-level"  storage logic.